_posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

```diff
@@ -0,0 +1,424 @@
+
+#### `torch`
+
+Installing `torch` is as straightforward as typing
+
+```{r}
+install.packages("torch")
+```
+
+This will detect whether you have CUDA installed, and either download the CPU or the GPU version of `libtorch`. Then, it will install the R package from CRAN. To make use of the very newest features, you can also install the development version from GitHub:
+
+```{r}
+devtools::install_github("mlverse/torch")
+```
+
+To quickly check the installation, and whether GPU support works fine (assuming that there *is* a GPU), create a tensor *on the CUDA device*:
+
+```{r}
+torch_tensor(1, device = "cuda")
+```
+
+    torch_tensor 
+     1
+    [ CUDAFloatType{1} ]
+
+If all our *hello torch* example did was running a network on, say, simulated data, we could stop here. As we'll do image classification, however, we need to install another package: `torchvision`.
+
+#### `torchvision`
+
+```{r}
+```
+
+Whereas `torch` is where tensors, network modules, and generic data loading functionality live, datatype-specific capabilities are -- or will be -- provided by dedicated packages. In general, these capabilities comprise three types of things: datasets, tools for pre-processing and data loading, and pre-trained models.
+
+As of this writing, PyTorch has dedicated libraries for three domain areas: vision, text, and audio. In R, we plan to proceed analogously -- "plan", because `torchtext` and `torchaudio` are still to be created. Right now, `torchvision` is all we need:
+
+```{r}
+install.packages("torchvision")
+```
+
+Note: As `torchvision` relies on [magick](https://cran.r-project.org/web/packages/magick/vignettes/intro.html), which itself needs [ImageMagick](https://imagemagick.org/) installed, you need to ensure the latter is available on your system. [The magick vignette](https://cran.r-project.org/web/packages/magick/vignettes/intro.html) has an excellent description on how to do this on different operating systems.
+
+And we're ready to load the data.
+
+## Data loading and pre-processing
+
+```{r}
+library(torch)
+library(torchvision)
+```
+
+The list of vision datasets bundled with PyTorch is long, and they're continually being added to `torchvision`.
+
+The one we need right now is available already, and it's -- MNIST? ... not quite: It's my favorite "MNIST dropin", [Kuzushiji-MNIST](https://github.com/rois-codh/kmnist) [@clanuwat2018deep]. Like other datasets explicitly created to replace MNIST, it has ten classes -- characters, in this case, depicted as grayscale images of resolution `28x28`.
+
+Here are the first 32 characters:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Kuzushiji MNIST."}
+knitr::include_graphics("images/kmnist.png")
+```
+
+#### Dataset
+
+The following code will download the data, separately for training and test sets.
+
+```{r}
+dir <- "."
+
+train_ds <- kmnist_dataset(
+  dir,
+  download = TRUE,
+  train = TRUE,
+  transform = function(x) {
+    x <- x$to(dtype = torch_float())/256
```

Comment:
there's a `transform_to_tensor` that takes a magick image, normalizes it into [0,1] and adds the channels dimension.

Reply:
right, we have that now... will change